serialization/dotnet/text/src/TextSerializationWriter.cs

```diff
@@ -0,0 +1,94 @@
+    /// <inheritdoc />
+    public void Dispose()
+    {
+        _stream?.Dispose();
+        GC.SuppressFinalize(this);
+    }
+    /// <inheritdoc />
+    public Stream GetSerializedContent() {
+        var result = writer.ToString();
+        _stream.Write(Encoding.UTF8.GetBytes(result));
+        writer.Clear();
+        return _stream;
+    }
+    /// <inheritdoc />
+    public void WriteAdditionalData(IDictionary<string, object> value) => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
+    /// <inheritdoc />
+    public void WriteBoolValue(string key, bool? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteByteArrayValue(string key, byte[] value) => WriteStringValue(key, value.Any() ? Convert.ToBase64String(value) : string.Empty);
+    /// <inheritdoc />
+    public void WriteByteValue(string key, byte? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteCollectionOfObjectValues<T>(string key, IEnumerable<T> values) where T : IParsable => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
+    /// <inheritdoc />
+    public void WriteCollectionOfPrimitiveValues<T>(string key, IEnumerable<T> values) => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
+    /// <inheritdoc />
+    public void WriteDateTimeOffsetValue(string key, DateTimeOffset? value) => WriteStringValue(key, value.HasValue ? value.Value.ToString() : null);
+    /// <inheritdoc />
+    public void WriteDateValue(string key, Date? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteDecimalValue(string key, decimal? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteDoubleValue(string key, double? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteFloatValue(string key, float? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteGuidValue(string key, Guid? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteIntValue(string key, int? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteLongValue(string key, long? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteNullValue(string key) => WriteStringValue(key, "null");
+    /// <inheritdoc />
+    public void WriteObjectValue<T>(string key, T value) where T : IParsable => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
+    /// <inheritdoc />
+    public void WriteSbyteValue(string key, sbyte? value) => WriteStringValue(key, value?.ToString());
+    /// <inheritdoc />
+    public void WriteStringValue(string key, string value)
+    {
+        if(!string.IsNullOrEmpty(key))
```

Comment:
I'm a big partisan of fail early, fail hard so what went wrong is obvious.
text/plain doesn't have any serialization standard for anything else than scalar values. Trying to write complex objects, collections, or just key values would result in an invalid payload... because there's no standard and therefore no way to know how the service wants it.
It the serialization lib silently tried to ignore the key, or worse, write it in "our own format", it would lead to payloads not reflecting the intended value the consumer had in mind. And without throwing, their only way to known something "fishy" is happening is to inspect data over the wire.  A whole lot more complex and insidious than seeing an exception in the logs.
Does that make sense?